google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/env/Emulator.java

```diff
@@ -0,0 +1,212 @@
+
+    try {
+      gcloudSdkPath = getGcloudSdkPath();
+    } catch (Exception e) {
+      throw new RuntimeException("Failed to get the gcloud SDK path. Is it installed?", e);
+    }
+
+    Path emulatorPath =
+        gcloudSdkPath.resolve(Paths.get("platform", "bigtable-emulator", "cbtemulator"));
+
+    if (!Files.exists(emulatorPath)) {
+      throw new RuntimeException(
+          "cbtemulator is not installed, please install with `gcloud components install bigtable`");
+    }
+
+    return new Emulator(emulatorPath);
+  }
+
+  private Emulator(Path executable) {
+    this.executable = executable;
+  }
+
+  void start() throws Exception {
+    int availablePort = getAvailablePort();
+
+    process = Runtime.getRuntime().exec(executable + " -port " + "" + availablePort);
+    pipeStreamToLog(process.getInputStream(), Level.INFO);
+    pipeStreamToLog(process.getErrorStream(), Level.WARNING);
+
+    waitForPort(availablePort);
+
+    channel = createChannel(availablePort);
+
+    tableAdminClient =
+        BigtableTableAdminClient.create(
+            configureClient(BigtableTableAdminSettings.newBuilder()).build());
+
+    dataClient =
+        BigtableDataClient.create(
+            configureClient(BigtableDataSettings.newBuilder().setInstanceName(INSTANCE_NAME))
+                .build());
+  }
+
+  void stop() throws Exception {
+    try {
+      dataClient.close();
+      tableAdminClient.close();
+      channel.shutdownNow();
+      channel.awaitTermination(1, TimeUnit.MINUTES);
+    } finally {
+      process.destroy();
+    }
+  }
+
+  BigtableDataClient getDataClient() {
+    return dataClient;
+  }
+
+  BigtableTableAdminClient getTableAdminClient() {
+    return tableAdminClient;
+  }
+
+  // <editor-fold desc="Helpers">
+  private static Path getGcloudSdkPath() throws Exception {
+    Process p = Runtime.getRuntime().exec("gcloud info --format=value(installation.sdk_root)");
+    pipeStreamToLog(p.getErrorStream(), Level.WARNING);
+
+    String sdkRoot = bufferOutput(p.getInputStream()).get(1, TimeUnit.MINUTES).trim();
+
+    if (p.waitFor() != 0) {
+      throw new RuntimeException("Failed to get sdk root, is gcloud sdk installed?");
+    }
+    return Paths.get(sdkRoot);
+  }
+
+  private static int getAvailablePort() {
+    try (ServerSocket serverSocket = new ServerSocket(0)) {
+      return serverSocket.getLocalPort();
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to find open port");
+    }
+  }
+
+  private void waitForPort(int port) throws InterruptedException, TimeoutException {
+    for (int i = 0; i < 100; i++) {
+      try (Socket ignored = new Socket("localhost", port)) {
+        return;
+      } catch (IOException e) {
+        Thread.sleep(200);
+      }
+    }
+
+    throw new TimeoutException("Timed out waiting for server to start");
+  }
+
+  private ManagedChannel createChannel(int port) {
+    return ManagedChannelBuilder.forAddress("localhost", port)
+        .usePlaintext(true)
```

Comment:
This is a bit problematic. I'm using an `@Experimental` grpc api. I'm wondering/hoping that since this is only used for tests its ok?

Reply:
I think that's ok actually - tests will never be involved in diamond dependency conflicts.

Reply:
I added a comment just in case :)